### PR TITLE
Button to add current location to UserConfiguration

### PR DESCRIPTION
**Pull request by tonylee80**
### Summary
🌟📝🧪

This pull request implements a bookmark feature that allows users to save and access their favorite pages in the app. It adds a new prop `displayAddBookmark` to the `AppShell` component to control the visibility of the add bookmark button in the `Navbar` component. It also adds a new component `BookmarkDialog` that is used to add a bookmark, and updates the tests and stories for the affected components.

*Note*
Only Users with UserConfiguration in their ProjectMembership can view the `+ bookmark` button at the bottom of the Appshell navbar.

> _`AppShell` is the gateway to your doom_
> _`displayAddBookmark` is the prop of gloom_
> _`BookmarkDialog` is the trap of fate_
> _You can't escape the bookmarks of hate_

### Walkthrough
*  Add a new prop `displayAddBookmark` to the `AppShell` component to control the visibility of the add bookmark button in the `Navbar` component ([link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-2d6eea6932e73947413887c8547a8756f34d08cb3c08ae70d70922e5c3f2102aL34-R39), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-0c5a7736d11c5b3df706061463ac8a39cf6c2c50bbae23e692398d6d0c4d9c42R64), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-f8990f8a0a0492bc2672c3a4ac95abd1b9d089490dc3efe59147235bc06dd629R14), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-f8990f8a0a0492bc2672c3a4ac95abd1b9d089490dc3efe59147235bc06dd629L49-R54), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-99025880609bfa25da082149ca3d96f1ab33f5ea74b570faca2a728ef1b53fbdR61), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-93b9f766bd8e36aeec9e0f0c338532c7747dc32719a0de0839970d22505d32aaR76), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-93b9f766bd8e36aeec9e0f0c338532c7747dc32719a0de0839970d22505d32aaL127-R151))
*  Implement the `BookmarkDialog` component to allow users to save bookmarks for their frequently visited pages ([link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-93b9f766bd8e36aeec9e0f0c338532c7747dc32719a0de0839970d22505d32aaL1-R5), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-93b9f766bd8e36aeec9e0f0c338532c7747dc32719a0de0839970d22505d32aaR82), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-93b9f766bd8e36aeec9e0f0c338532c7747dc32719a0de0839970d22505d32aaL127-R151), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-5c498f976bbc87ba3cfb734920c9383e0b29c636007f70f94e04e56c09ca26b7R1-R212), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-ee7d0b24d96e07682409c84e3288d3ec2c2ca3efee8ea39579d8d66ee7fb9bacR1-R83))
*  Update the `Navbar` component to render the add bookmark button and the `BookmarkDialog` component conditionally based on the `displayAddBookmark` prop and the `bookmarkDialogVisible` state ([link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-93b9f766bd8e36aeec9e0f0c338532c7747dc32719a0de0839970d22505d32aaR82), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-93b9f766bd8e36aeec9e0f0c338532c7747dc32719a0de0839970d22505d32aaL127-R151))
*  Improve the layout and appearance of the `CodeInput` component in the `Navbar` component by adding some margin and spacing ([link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-93b9f766bd8e36aeec9e0f0c338532c7747dc32719a0de0839970d22505d32aaL96-R120))
*  Write unit tests for the `Navbar` component and the `BookmarkDialog` component to ensure the quality and correctness of the bookmark feature ([link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-99025880609bfa25da082149ca3d96f1ab33f5ea74b570faca2a728ef1b53fbdR14-R36), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-99025880609bfa25da082149ca3d96f1ab33f5ea74b570faca2a728ef1b53fbdR185-R249), [link](https://github.com/medplum/medplum/pull/1927/files?diff=unified&w=0#diff-5c498f976bbc87ba3cfb734920c9383e0b29c636007f70f94e04e56c09ca26b7R1-R212))

###Screenshots

Screenshot 2023-04-29 at 7 48 58 PM

image

image

### CODE REVIEW

File: packages/react/src/AppShell/Navbar.tsx [hunk 1]
```diff
@@ -92,41 +105,118 @@ export function Navbar(props: NavbarProps): JSX.Element {
...
+            family: value,
+            given: ['John'],
+          },
+        ],
+      })
+      .then((response) => {
+        console.log(response);
+      })
+      .catch((error) => {
+        console.log(error);
```
Comment (tonylee80) [inline packages/react/src/AppShell/Navbar.tsx:125]: todo: notification for success and errors

File: packages/react/src/AppShell/Navbar.tsx [hunk 2]
```diff
@@ -92,41 +105,118 @@ export function Navbar(props: NavbarProps): JSX.Element {
     }
   }
 
+  function addBookmark(value: string): void {
+    console.log('adding bookmark', value);
+    medplum
+      .updateResource({
+        resourceType: 'Patient',
```
Comment (tonylee80) [inline packages/react/src/AppShell/Navbar.tsx:112]: update user projectmembership userconfiguration

File: packages/react/src/AppShell/Navbar.tsx (hunk 1 above)
Comment (tonylee80) [reply packages/react/src/AppShell/Navbar.tsx:125]: non `admin` users will not be able to modify userconfiguration

File: packages/react/src/AppShell/Navbar.tsx [hunk 3]
```diff
@@ -92,41 +105,118 @@ export function Navbar(props: NavbarProps): JSX.Element {
...
+          setState({
+            ...stateRef.current,
+            addBookmarkVisible: false,
+          });
+        }}
+      />
+    </>
+  );
+}
+interface AddBookmarkModalProps {
```
Comment (tonylee80) [inline packages/react/src/AppShell/Navbar.tsx:193]: todo: move to separate file

File: packages/core/src/client.ts [hunk 4]
```diff
@@ -2056,7 +2056,7 @@ export class MedplumClient extends EventTarget {
     this.storage.setObject('logins', logins);
   }
 
-  private async refreshProfile(): Promise<ProfileResource | undefined> {
+  async refreshProfile(): Promise<ProfileResource | undefined> {
```
Comment (tonylee80) [inline packages/core/src/client.ts:2059]: this helped refresh the AppShell leftnav, but open to find another way to refresh the UserConfiguration

Comment (codyebberson) [reply packages/core/src/client.ts:2059]: The trick is to get `<MedplumProvider>` to re-render.

Here is where `<MedplumProvider>` listens for change events: https://github.com/medplum/medplum/blob/main/packages/react/src/MedplumProvider/MedplumProvider.tsx#L37-L47

Here is background reading on React Context: https://legacy.reactjs.org/docs/context.html

File: packages/react/src/BookmarkDialog/BookmarkDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  return (
+    <Modal
+      title="Add Bookmark"
+      closeButtonProps={{ 'aria-label': 'Close' }}
+      opened={props.visible}
+      onClose={props.onCancel}
+    >
+      <Form onSubmit={submitHandler}>
```
Comment (codyebberson) [inline packages/react/src/BookmarkDialog/BookmarkDialog.tsx:48]: In most of the app forms, we use the Mantine `<Stack>` component for better vertical spacing.

https://mantine.dev/core/stack/

```tsx
<Form ...>
  <Stack>
    <TextInput ... />
    <TextInput ... />
    <TextInput ... />
  </Stack>
</Form>
```

**Review by codyebberson — CHANGES_REQUESTED** [review]
Nice, this is going to be super useful.

See comments on "Bookmark" vs "BookMark"

Aesthetically, I played with some tweaks to the button and icon styles, for your consideration: https://github.com/medplum/medplum/pull/1943

File: packages/react/src/AppShell/AppShell.tsx [hunk 6]
```diff
@@ -46,7 +47,11 @@ export function AppShell(props: AppShellProps): JSX.Element {
       padding={0}
       fixed={true}
       header={profile && <Header logo={props.logo} version={props.version} navbarToggle={toggleNavbar} />}
-      navbar={profile && navbarOpen ? <Navbar menus={props.menus} closeNavbar={closeNavbar} /> : undefined}
+      navbar={
+        profile && navbarOpen ? (
+          <Navbar menus={props.menus} closeNavbar={closeNavbar} displayAddBookMark={props.displayAddBookmark} />
```
[the remark this answers is not in the document]
Comment (tonylee80) [reply packages/react/src/AppShell/AppShell.tsx:52]: Thanks for catching the Typo!

**Comment by tonylee80** [comment]
> 

Thanks for the PR! I've went ahead and merged the UI changes.

**Comment by tonylee80** [comment]
> 



> Nice, this is going to be super useful.
> 
> See comments on "Bookmark" vs "BookMark"
> 
> Aesthetically, I played with some tweaks to the button and icon styles, for your consideration: #1943

Thanks for the PR for the UI changes! I've went ahead and merged #1943 .

**Review by codyebberson — APPROVED** [review]
lgtm 👍

**Pull request merged**